src/endpoints/voting-history/data/repositories/user-voting-history.repository-impl.ts

```diff
@@ -0,0 +1,126 @@
+import { Collection, CollectionSource, FindOptions, Mapper, MongoSource, RepositoryImpl, Result } from '@alien-worlds/api-core';
+import { UserVotingHistoryDocument, VoteAction } from './../dtos/user-voting-history.dto';
+
+import { UserVote } from '../../domain/entities/user-vote';
+import { VotingHistoryQueryModel } from '../../domain/models/voting-history.query-model';
+
+export class UserVotingHistoryRepositoryImpl extends RepositoryImpl<
+  UserVote,
+  UserVotingHistoryDocument
+> {
+  private actions: Collection;
+  private contractRows: Collection;
+
+  constructor(
+    source: CollectionSource<UserVotingHistoryDocument>,
+    mapper: Mapper<UserVote, UserVotingHistoryDocument>,
+    mongoSource: MongoSource
+  ) {
+    super(source, mapper);
+    this.actions = mongoSource.database.collection('actions');
+    this.contractRows = mongoSource.database.collection('contract_rows');
+  }
+
+  async find(
+    model: VotingHistoryQueryModel
+  ): Promise<Result<UserVote[], Error>> {
+    const output: UserVote[] = [];
+
+    const { filter, options } = model.toQueryParams();
+    const fetchLastEntryFromPrevPage = options.skip > 0;
+
+    const actionsRes = await (await this.actions.find(filter, options)).toArray();
+
+    /*
+     * fetch previous entry to calculate Voting Action 
+     * of the first entry in response especially when we do pagination
+     */
+    let lastEntry;
+    if (fetchLastEntryFromPrevPage) {
+      lastEntry = await this.actions.findOne(filter, {
+        ...options,
+        skip: options.skip - 1,
+        limit: 1,
+      });
+    }
+
+    for (let i = 0; i < actionsRes.length; i++) {
+      const vote = actionsRes[i];
+      const actionData = vote.action.data;
+
+      let prevVotedCandidates: string[] = [];
+      if (i == 0 && fetchLastEntryFromPrevPage) {
```

Comment:
It would be good to put some more comments in this method to explain these conditions. It will be helpful to get context when we get back to this code after some time